### ISSUE TITLE
auth: Support sign in for Facebook limited login mode

### ISSUE THREAD

**Issue by MegaMaddin** [firebase/firebase-ios-sdk#8048]
### Feature proposal

* Firebase Component: Auth

Facebook recently added the ability to sign in using a limited data and tracking mode called [limited login](https://developers.facebook.com/blog/post/2021/01/19/facebook-login-updates-new-limited-data-mode/). The benefit of that is that you are able to tell Facebook in advance to not track the user and that the data you can get out of Facebook is limited to a couple attributes (sufficient for using it with Firebase Auth). Since the token returned is a [different one](https://developers.facebook.com/docs/facebook-login/limited-login/) you would get usually when signing in with Facebook, I guess this also requires changes in the Firebase Auth backend.

Please let me know if you need further information.

Best,
Martin

**Comment by yycking** [comment]
How to let Firebase Auth support [Facebook OprnID](https://developers.facebook.com/docs/facebook-login/limited-login/token) ?

**Comment by weixifan** [comment]
@MegaMaddin Thank you for filing this feature request! We are unable to promise any timeline for this, but if others also have this request, adding a +1 on this issue can help us prioritize adding this to the roadmap.

Internal tracking bug: b/188589295.

**Comment by abidmasood** [comment]
+1

**Comment by yycking** [comment]
+1

**Comment by chuan7848** [comment]
+1

**Comment by patrickmegazebra** [comment]
+1

**Comment by mikehardy** [comment]
I think what the firebase-ios-sdk team member means by `+1` is to add a reaction thumbs up above :point_up: 

@craig1123 - this looks like the natural upstream home of [the issue you raised in react-native-firebase](https://github.com/invertase/react-native-firebase/issues/5441)

**Comment by mikehardy** [comment]
In the absence of a native solution, here's a (painful) idea:

I can imagine a difficult workarounds in the meantime, involving self-validation of the Facebook Limited Login AuthenticationToken https://developers.facebook.com/docs/facebook-login/limited-login/token/validating followed by minting your own Firebase Auth Token https://firebase.google.com/docs/auth/android/custom-auth
I think the flow would be:
• app implements Limited Login on iOS via FBSDK
• app sends that JWT to custom firebase cloud function
• custom firebase cloud function validates JWT and if valid mints custom Auth token using admin sdk and returns it
• app uses this custom auth token for firebase access
All sorts of implementation details - nonce-handling while performing auth and minting tokens, token refresh how?, user resolution / de-duplication from "firebase+facebook native" existing users to this new limited login regime, etc

I'm posting this idea mostly as the sort-of "issue ombudsman" for react-native-firebase, I have not personally done anything like this so it is not informed by any sort of experience.

Could anything like this work though?

**Comment by marexv** [comment]
I was able to implement flow described by @mikehardy. I am able to log in users and it works OK. Nonce handling is a bit of an issue, but token refresh should be fine? Facebook token is used only once, after that you use firebase token and things are handled "as is". Biggest issue so far is allowing a single user, to change between two modes as there is no way to manually set login provider in firebase (i.e. limited login won't have a provider, and "ordinary" login will have facebook set as provider). You could achieve this interoperability by exchanging "ordinary" access token for firebase token on backend, but then it is a new endpoint and more validation logic, and it is in fact easier to handle these edge cases manually. Really looking forward to a rn-firebase solution for this :)

**Comment by mikehardy** [comment]
Wow! Champion effort, I'm happy the sketch was useful at all and kind of amazed it worked. Just to note that prior to any react-native-firebase solution it will need real support here (firebase-ios-sdk) first, then it can come up a level to react-native-firebase where we just wrap things.

Any chance you could post a gist or a repo with the chunks of code required to implement this? That could help - might even help the firebase-ios-sdk team here move things forward

**Comment by marexv** [comment]
Sure, would you like to the see backend code too? Its .net core 3.1

**Comment by mikehardy** [comment]
I think any functioning code is useful as long as you can publish. Just so it's clear exactly what data elements are inspected / transformed / created / passed-on etc right?

**Comment by marexv** [comment]
I have sent you an email with a gist link, if you have some time, please take a look at it.

**Comment by rosalyntan** [comment]
Hi folks, Facebook Limited Login is now supported for Firebase Auth. Please see the [docs](https://firebase.google.com/docs/auth/ios/facebook-login#implement_facebook_limited_login) for usage guides.

In addition, Facebook Limited Login support was added to FirebaseUI iOS in firebase/FirebaseUI-iOS#995 and is available in the latest released version.

**Comment by rosalyntan** [comment]
Hi @skyblue168, thanks for flagging this!

Seems like [Facebook recently updated the `iss` field in their ID token payload](https://developers.facebook.com/support/bugs/3019917938239359), although it looks like the new value is used only if you log in with the v12 Facebook iOS SDK.

I've forwarded this on to our backend team and will update here once we have a fix. For an immediate workaround, you can use FBSDKLoginKit v11 instead.

**Comment by skyblue168** [comment]
> Hi @skyblue168, thanks for flagging this!
> 
> Seems like [Facebook recently updated the `iss` field in their ID token payload](https://developers.facebook.com/support/bugs/3019917938239359), although it looks like the new value is used only if you log in with the v12 Facebook iOS SDK.
> 
> I've forwarded this on to our backend team and will update here once we have a fix. For an immediate workaround, you can use FBSDKLoginKit v11 instead.

OK, I will try it. thanks.

**Comment by rosalyntan** [comment]
Hi @skyblue168, the backend fix should now be fully rolled out, so you can now use the v12 Facebook iOS SDK.

Closing out this issue, as this deals with the initial feature request for Facebook Limited Login support. For any future issues, please open a new issue.

Thanks!
